File: generated_code/cpp/solution_2681_3.cpp
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long putMarbles(vector<int>& weights, int k) {
        // Core Idea: Splitting the array into k bags requires choosing k - 1 partition points.
        // If we split between index i and i + 1, the right boundary of the current bag is weights[i]
        // and the left boundary of the next bag is weights[i + 1]. Thus, every split contributes 
        // weights[i] + weights[i + 1] to the total score. 
        // The first element weights[0] and the last element weights[n - 1] are always part of the 
        // score regardless of how we split.
        // Therefore, finding the max/min total score is equivalent to finding the sum of the 
        // largest/smallest k - 1 adjacent pair sums.

        int n = weights.size();
        if (k == 1 || n == k) {
            return 0; // Only one valid distribution, so max_score - min_score = 0
        }

        // Generate all possible adjacent pair sums
        vector<long long> pair_sums(n - 1);
        for (int i = 0; i < n - 1; ++i) {
            pair_sums[i] = static_cast<long long>(weights[i]) + weights[i + 1];
        }

        // To optimize, we don't need a full sort. We can use std::nth_element 
        // to find the smallest and largest k - 1 elements in O(N) time.
        int num_splits = k - 1;

        // Find the sum of the smallest k - 1 elements
        nth_element(pair_sums.begin(), pair_sums.begin() + num_splits, pair_sums.end());
        long long min_splits_sum = 0;
        for (int i = 0; i < num_splits; ++i) {
            min_splits_sum += pair_sums[i];
        }

        // Find the sum of the largest k - 1 elements
        nth_element(pair_sums.begin(), pair_sums.begin() + (n - 1 - num_splits), pair_sums.end());
        long long max_splits_sum = 0;
        for (int i = n - 1 - num_splits; i < n - 1; ++i) {
            max_splits_sum += pair_sums[i];
        }

        // The difference in total scores is exactly the difference between the partition choices,
        // because weights[0] and weights[n - 1] cancel out.
        return max_splits_sum - min_splits_sum;
    }
};

File: generated_code/python/solution_2681_3.py
from typing import List

class Solution:
    def putMarbles(self, weights: List[int], k: int) -> int:
        # The problem asks for the difference between the maximum and minimum score
        # when partitioning the array into k contiguous subarrays (bags).
        # Notice that for any valid partition, the first element (weights[0]) and the
        # last element (weights[-1]) will always contribute to the total score, regardless
        # of how we partition the remaining elements.
        #
        # Each split between two bags at index i and i+1 adds weights[i] + weights[i+1]
        # to the score because it forms the end of one bag and the start of the next.
        # To make k bags, we need to make exactly k - 1 splits.
        #
        # Therefore, the score maximization/minimization depends entirely on choosing
        # the largest or smallest k - 1 adjacent pairs (weights[i] + weights[i+1]).
        
        n = len(weights)
        if k == 1 or k == n:
            return 0
        
        # Calculate the costs of all possible splits
        pair_weights = [weights[i] + weights[i + 1] for i in range(n - 1)]
        
        # Sort the pairs to easily pick the smallest and largest splits
        pair_weights.sort()
        
        # The number of splits needed is k - 1
        num_splits = k - 1
        
        # The maximum score uses the largest k - 1 pairs.
        # The minimum score uses the smallest k - 1 pairs.
        # When finding the difference, the constant terms (weights[0] + weights[-1]) cancel out.
        min_score_splits = sum(pair_weights[:num_splits])
        max_score_splits = sum(pair_weights[len(pair_weights) - num_splits:])
        
        return max_score_splits - min_score_splits